app/src/token_verifier.rs

```diff
@@ -0,0 +1,96 @@
+use crate::{
+    blake2b_hasher::{Blake2bHash, Blake2bHasher, BLAKE2B_HASH_SIZE},
+    error_code::ErrorCode,
+    handler::TOKEN_METADATA_SIZE,
+};
+
+// b3380866c595544781e9da0ccd79399de8878abfb0bf40545b57a287387d419d
+const TOKEN_MERKLE_ROOT: Blake2bHash = [
+    0xb3, 0x38, 0x08, 0x66, 0xc5, 0x95, 0x54, 0x47, 0x81, 0xe9, 0xda, 0x0c, 0xcd, 0x79, 0x39, 0x9d,
+    0xe8, 0x87, 0x8a, 0xbf, 0xb0, 0xbf, 0x40, 0x54, 0x5b, 0x57, 0xa2, 0x87, 0x38, 0x7d, 0x41, 0x9d,
+];
+const PROOF_PREFIX_LENGTH: usize = 2;
+
+// `TokenVerifier` is a streaming token proof verifier that receives proof data and calculates the hash
+// After receiving all the proof data, it compares the hash with the `TOKEN_MERKLE_ROOT` to verify if the token is valid
+#[derive(Default, Copy, Clone)]
+pub struct TokenVerifier {
+    remaining_proof_size: usize,
+    hash: Blake2bHash,
+}
+
+pub fn hash_pair(a: &[u8], b: &[u8]) -> Result<Blake2bHash, ErrorCode> {
+    assert!(a.len() == BLAKE2B_HASH_SIZE && b.len() == BLAKE2B_HASH_SIZE);
+    let mut hasher = Blake2bHasher::new();
+    if a < b {
+        hasher.update(a)?;
+        hasher.update(b)?;
+    } else {
+        hasher.update(b)?;
+        hasher.update(a)?;
+    }
+    hasher.finalize()
+}
+
+impl TokenVerifier {
+    pub fn new(data: &[u8]) -> Result<TokenVerifier, ErrorCode> {
+        let prefix_length = TOKEN_METADATA_SIZE + PROOF_PREFIX_LENGTH;
+        if data.len() < prefix_length {
+            return Err(ErrorCode::BadLen);
+        }
+
+        let encoded_token = &data[..TOKEN_METADATA_SIZE];
+        let proof_size =
+            ((data[TOKEN_METADATA_SIZE] as usize) << 8) | (data[TOKEN_METADATA_SIZE + 1] as usize);
+        check_proof_size(proof_size)?;
+
+        let proof = &data[prefix_length..];
+        check_proof_size(proof.len())?;
+
+        let mut verifier = TokenVerifier {
+            remaining_proof_size: proof_size,
+            hash: Blake2bHasher::hash(encoded_token)?,
+        };
+        verifier.update(proof)?;
```

Comment:
We could also use `verify.on_proof()` and remove the `check_proof_size` above. Then we could remove function `update`